src/lib/SaferNames/Syntax.hs

```diff
@@ -436,147 +340,249 @@ instance Semigroup (SourceNameMap n) where
-
-instance SubstE e => HasNamesE (SubstVal e) where traverseNamesE = traverseNamesFromSubstE
-instance SubstE e => SubstE (SubstVal e) where
-  traverseCoreE = undefined
-
-instance SubstE e => HasNamesE (WithArrow e) where
-  traverseNamesE = traverseNamesFromSubstE
-instance SubstE e => SubstE (WithArrow e) where
-  traverseCoreE s t (WithArrow arr e) =
-    WithArrow <$> traverse (tce s t) arr <*> tce s t e
-
-instance HasNamesE Block where traverseNamesE = traverseNamesFromSubstE
-instance SubstE Block where
-  traverseCoreE s t (Block decls result) =
-    traverseCoreE s t (Abs decls result) >>= \case
-      Abs decls' result' -> return $ Block decls' result'
+instance InjectableE DataDef where
+  injectionProofE = undefined
 
-instance HasNamesB Decl where
-  traverseNamesB = traverseNamesFromSubstB
-  boundScope = undefined
-instance SubstB Decl where
-  traverseCoreB s t (Let ann b expr) = do
-    expr' <- traverseCoreE s t expr
-    traverseCoreB s t b >>= \case
-      FreshBinder ext b' renamer ->
-        return $ FreshBinder ext (Let ann b' expr') renamer
-  asBindingsFrag = undefined
-
-instance HasNamesB DataConRefBinding where
-  traverseNamesB = undefined
-  boundScope = undefined
-
-instance SubstB b => SubstB (Nest b) where
-  traverseCoreB s t nest = case nest of
-    Empty -> do return $ FreshBinder emptyFreshExt Empty emptyEnv
-    Nest b rest ->
-      traverseCoreB s t (NestPair b rest) >>= \case
-        FreshBinder ext (NestPair b' rest') renamer ->
-          return $ FreshBinder ext (Nest b' rest') renamer
-  asBindingsFrag = undefined
-
-instance (SubstB b1, SubstB b2) => SubstB (NestPair b1 b2) where
-  traverseCoreB = undefined
-  asBindingsFrag = undefined
-
-instance HasNamesE NamedDataDef where traverseNamesE = traverseNamesFromSubstE
-instance SubstE NamedDataDef where
-  traverseCoreE = undefined
-
-instance HasNamesE DataDef where traverseNamesE = traverseNamesFromSubstE
-instance SubstE DataDef where
-  traverseCoreE = undefined
-
-instance HasNamesE Atom where traverseNamesE = traverseNamesFromSubstE
-instance SubstE Atom where
-  traverseCoreE s t atom = case atom of
-    Var (AnnVar v ann) ->
-      lookupNameTraversal t v >>= \case
-        Left substVal -> case substVal of
-          SubstVal x -> return x
-          Rename v' -> do
-            ann' <- traverseCoreE s t ann
-            return $ Var $ AnnVar v' ann'
-        Right v' -> do
-          ann' <- traverseCoreE s t ann
-          return $ Var $ AnnVar v' ann'
-    Lam ab -> Lam <$> tce s t ab
-    Pi  ab -> Pi  <$> tce s t ab
+instance HasNamesE DataDef where
+  traverseNamesE = undefined
+
+instance SubstE AtomSubstVal DataDef where
+  substE = undefined
+
+instance InjectableE DataConDef where
+  injectionProofE = undefined
+
+instance HasNamesE DataConDef where
+  traverseNamesE = undefined
+
+instance SubstE AtomSubstVal DataConDef where
+  substE = undefined
+
+instance InjectableE Atom where
+  injectionProofE = undefined
```

Comment:
Are you planning to fill those in before merging?

Reply:
Sure. Happy to.